### ISSUE TITLE
External assets

### ISSUE THREAD

**Issue by zdne** [apiaryio/api-blueprint#20]
Add the possibility to use [Markdown links](http://daringfireball.net/projects/markdown/syntax#link) wherever an asset is expected (e.g. a payload body or schema).

**Comment by stroml** [comment]
You can use [gulp-file-include](https://github.com/coderhaoxin/gulp-file-include) and create a gulp task to watch your json and build the api. The only downside is that you have to have an api.source.md to generate your api.md, but IMO it's a small price to pay.

**Comment by zdne** [comment]
Also for what it's worth checkout @jamesramsay 's Hercule – https://github.com/jamesramsay/hercule

**Comment by jamesramsay** [comment]
Thanks @zdne for the mention. I had similar challenges when writing API documentation and we have been using [Hercule](https://github.com/jamesramsay/hercule) internally for the last 3 months.

Hercule uses 


We have a number of API contexts for different types of API consumers that have nearly identical resources and we were keen to reduce duplication between the different documents.

**Comment by zdne** [comment]
@jamesramsay 

I am just wondering why to use `{{link}}` syntax instead of Markdown's very own `[<placeholder>](link)` ? as discussed above. For example: 

```markdown
=[gist](gist.apib)
```

or 

```
:[gist](gist.apib)
```

it would work similarly, it would render reasonably as markdown, so you would be able to navigate through links, and you could use much richer syntax such as `[placeholder][]` with `[placeholder] link options...

**Comment by jamesramsay** [comment]
@zdne this is a really great suggestion! Too bad I didn't see this thread 6 months ago.

The `{{link}}` syntax was based on [MultiMarkdown 4's](http://fletcher.github.io/MultiMarkdown-4/transclusion.html) transclusion syntax. It did the trick and I wasn't inventing a new syntax.

At the time I also looked at Marked app (`<<[link]`) and a couple of other Markdown compilers.

With the advantage of being able to click through the links when viewing the source documentation in Github and improved consistency with the feel of Markdown I've updated hercule to use `:[gist](gist.apib)`. If anyone would like to take it for a spin, it's on NPM.

**Comment by reorx** [comment]
Hey guys, I am using API Blueprint with [aglio](https://github.com/danielgtaylor/aglio) for our internal project, and since API Blueprint doesn't support file including currently, aglio implement one with the syntax `<!-- include(filename.json) -->`, thanks to this feature we can split request/resposne JSON data from API document clearly, but when we organizing the seperated JSON data files, we find it still lack of a proper way to reuse the same part, for example, we have a `user_model.json`:

```json
{
    "username": "alice",
    "comments": [
        {
            "id": 1,
            "content": "hello"
        }
    ]
}
```

and a `comment_model.json`:

```json
{
    "id": 1,
    "content": "hello"
}
```

the comment part in `user_model.json` is and should always be the same as `comment_model.json`, if there is a mechanism that allows JSON file to include each other, it would be very easy to maintain their consistency, so that in case the commend model is changed, we don't need to change it everywhere.

I consider this for some time and try making a syntax to support JSON file including, here is my attempt with a Python implementation: https://github.com/reorx/json_include

The syntax would be like, in previous case:

```json
{
    "username": "alice",
    "comments": [
        {
            "...": "include(commend_model.json)"
        }
    ]
}
```

using `{ "...": "include(commend_model.json)"}` to represent a including to `commend_model.json` file.

Its not a special case for API Blueprint, but something rather general, to try to think of JSON itself with the include capability, which could also be useful in API Blueprint.

Please take a look and give some suggestions, currently this [json-include](https://github.com/reorx/json_include) is working fine in some of my projects, any thoughts about it would be very appreciated, since I can't decide whether it is good and meaningful alone by myself :)

**Comment by zdne** [comment]
Hey @reorx ! Thanks for sharing the idea. Recently we have launched (beta) of [MSON](https://github.com/apiaryio/mson) – a syntax for describing data (not only) in API Blueprint. The idea is, you describe your data once and then ask for serialization into different formats. 


Using your examples it works like this:
        
```markdown
# Data Structures

## User 
- username: alice
- comments (array[Comment])

## Comment
- id: 1
- content: hello
```
        
and then you can ask to render it as JSON:

```markdown
# Resource [/r]
## Retrieve [GET]
+ Response (application/json)
    + Attributes (User)

```
        
What do you think? Will this help you?

**Comment by jonathancrosmer** [comment]
Is this feature anywhere in the pipeline?  It would be very nice for us.
We use Blueprint to describe methods, but have our JSON schema hosted separately.  We have to include the schema as links in the method descriptions, since it isn't supported under the request/response element currently. 
Actually the most ideal thing would be if we could "include" the JSON schema using a link under the request/response element, but it would have to be lazily loaded, to avoid doing hundreds of includes when the documentation page is loaded.

**Comment by lpbm** [comment]
I think that the fact you can't include external references for requests/response body examples and definition schema files is one of the _very_ big downsides that Blueprint has in respect to other API definition languages.

**Comment by moon0326** [comment]
+1

**Comment by zdne** [comment]
@jonathancrosmer @lpbm @moon0326 

check https://github.com/jamesramsay/hercule especially 

https://github.com/jamesramsay/hercule/blob/master/examples/api-blueprint/gist-fox.apib

and 

https://github.com/jamesramsay/hercule/blob/master/examples/api-blueprint/blueprint/gists.md

would this work for you?

**Comment by moon0326** [comment]
@zdne I'm not following.

**Comment by lpbm** [comment]
@zdne: yes, as long as it allows inclusion of any type of external resource, be it .md/.apib or xml, json that's good. 
 
Is it something Apiary will implement, or is your message an advise for us to use hercule before committing to apiary?

**Comment by zdne** [comment]
Back ago I have prepared a list of solutions for splitting blueprint into multiple files (and using exteral assets):  https://gist.github.com/zdne/137396a1c2d45f75b306

@jamesramsay put together the awesome Hercule tool which implements the syntax I would like to use for splitting blueprint into multiple parts. 

As for the external assets. Hercule works there as well albeit I am not sure if this should be the final syntax and suggested use inside JSON blobs especially through the prism of [MSON](https://github.com/apiaryio/mson). 

Nevertheless if you do not rely on editing inside Apiary you can use Hercule today to get most of this functionality. 

> Is it something Apiary will implement? 

We want to implement some transclusion based on the `:[](url)` syntax, yes. The catch here is the Apiary editor and implementation of a "filesystem".  Until then you would have to use Hercule and push concatenated file into Apiary.

Does this clarify?

**Comment by lpbm** [comment]
> The catch here is the Apiary editor and implementation of a "filesystem"

Maybe you can circumvent that by using URLs?

**Comment by zdne** [comment]
For fetching the documents perhaps yes but not for editing.

**Comment by jonathancrosmer** [comment]
Currently we have an Apiary blueprint stored in our git repository, which is connected to Apiary.  The blueprint has many links to an internal server that hosts our JSON schema.
I suppose with this solution we would have an original Markdown file, then use a Git hook or something to run the transclusion and commit the result as the sync'd blueprint.  Sounds like it could work; a bit of a hassle to set up the transclusion process, but it would be slick if it works, so I will look into it.  Thanks!

**Comment by jamesramsay** [comment]
@lpbm Hercule supports both local file (relative file paths) and remote file (URL) transclusion, and although labelled as a 'Markdown transclusion' should support any text file (.md, .apib, .json, .xml etc). One use case that might be similar to yours that some people are using Hercule for is transcluding JSON Schema generated by their application via HTTP, and also local file transclusion to increase the DRY-ness of their documentation.

@jonathancrosmer spot on - we do something similar using Jenkins on our internal CI environment. At Adslot, whenever a commit is pushed to any branch, along with unit tests we compile the API Blueprint with Hercule, and then pass it into dredd. If all is passing, as part of the deploy task, we can optionally push to Apiary. We originally pushed directly to Apiary as soon as changes were made to the API docs, but we found this sometimes caused problems when we were working on larger releases with feature switches and such, thus we now deploy alongside the application.

**Comment by bartfeenstra** [comment]
What about just allowing plain URIs to be specified? The renderer can decide whether to render the URI to a link or include its contents in the documentation output. This also saves us the trouble from having to deal with different Markdown flavors.

**Comment by zdne** [comment]
> What about just allowing plain URIs to be specified?

That should be the case. `[label](URI)` should be used for transclustion and also render well. The difference is we need to mark if for the transclustion tool hence `:[label](URI)`

**Comment by smil2k** [comment]
Hi,

Is there any news on this? 
I think schema and examples should be separated from the main body of the api documentation, so they can be used by the server and/or client for validation and unit tests. Links are ok.

Cheers,

Tamas

**Comment by clucasalcantara** [comment]
Hello everyone, I'm sorry for getting here late and reopening and digging old stuff. If this isn't the right place please forgive me. But, doing a research to develop blueprint mocks I've found this discussion and this is the closest that I've could find until now.

I'm struggling to split my responses by model :(. Isn't it possible to referer a model into a different folder? 
I would like to have inside my blueprints folder all my requests, and, for example into a separated models folder all my entities models (MSON/MD/JSON/whatever files, that whose fit) to reuse modes into different responses easily. Anyone have success doing this?
